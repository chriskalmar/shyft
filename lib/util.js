
import DataType from './datatype/DataType';
import Entity from './entity/Entity';


export const passOrThrow = (condition, messageFn) => {
  // providing the error message as a callback return massively improves code coverage reports.
  // the message function is only evaluated if the condition fails, which will show up correctly
  // in the coverage reports.
  if (typeof messageFn !== 'function') {
    throw new Error('passOrThrow() expects messageFn to be a function')
  }
  if (!condition) {
    throw new Error(messageFn())
  }
}


export const isDataType = (obj) => {
  return (obj instanceof DataType)
}


export const isEntity = (obj) => {
  return (obj instanceof Entity)
}


export const resolveFunctionMap = (functionOrMap) => {
  return typeof functionOrMap === 'function'
    ? functionOrMap()
    : functionOrMap
}


export const isMap = (map, nonEmpty) => {
  return map !== null &&
    typeof map === 'object' &&
    Array.isArray(map) === false &&
    (!nonEmpty || (nonEmpty && Object.keys(map).length > 0 ))
}


export const isFunction = (fn) => {
  return typeof fn === 'function'
}



export const isArray = (set, nonEmpty) => {
  return set !== null &&
    typeof map === 'array' &&
    Array.isArray(set) === true &&
    (!nonEmpty || (nonEmpty && set.length > 0 ))
}

