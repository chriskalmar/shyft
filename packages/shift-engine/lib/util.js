import casual from 'casual';
import _ from 'lodash';



export const deleteUndefinedProps = (obj) =>
  Object.keys(obj).forEach(key => obj[key] === undefined && delete obj[key])



export const passOrThrow = (condition, messageFn) => {
  // providing the error message as a callback return massively improves code coverage reports.
  // the message function is only evaluated if the condition fails, which will show up correctly
  // in the coverage reports.
  if (typeof messageFn !== 'function') {
    throw new Error('passOrThrow() expects messageFn to be a function')
  }
  if (!condition) {
    throw new Error(messageFn())
  }
}


export const resolveFunctionMap = (functionOrMap) => {
  return typeof functionOrMap === 'function'
    ? functionOrMap()
    : { ...functionOrMap }
}


export const isMap = (map, nonEmpty) => {
  return map !== null &&
    typeof map === 'object' &&
    Array.isArray(map) === false &&
    (!nonEmpty || (nonEmpty && Object.keys(map).length > 0 ))
}


export const isFunction = (fn) => {
  return typeof fn === 'function'
}



export const isArray = (set, nonEmpty) => {
  return set !== null &&
    Array.isArray(set) === true &&
    (!nonEmpty || (nonEmpty && set.length > 0 ))
}


export const isString = (str) => (typeof str === 'string')


export const mergeMaps = (first, second) => {
  if (!isMap(first) || !isMap(second)) {
    throw new Error('mergeMaps() expects 2 maps for a merge to work')
  }

  return { ...first, ...second }
}



export const mapOverProperties = (object, iteratee) => {

  passOrThrow(
    isMap(object),
    () => 'Provided object is not a map'
  )

  passOrThrow(
    isFunction(iteratee),
    () => 'Provided iteratee is not a function'
  )


  const keys = Object.keys(object);

  keys.forEach((key) => {
    iteratee(object[ key ], key)
  })
}



export const sortDataByKeys = (keys, data, keyProperty='id') => {
  const map = {}
  const result = []
  const order = {}

  if (!keys || ( isArray(keys) && keys.length === 0)) {
    return []
  }

  if (!data || ( isArray(data) && data.length === 0)) {
    return keys.map(() => null)
  }


  for (let d = 0; d < data.length; d++) {
    const row = data[ d ]
    const id = row[ keyProperty ]
    const found = map[id]

    if (!found) {
      map[id] = row
    }
    else if (found instanceof Array) {
      found.push(row)
    }
    else {
      map[id] = [ found, row ];
    }
  }


  for (let k = 0; k < keys.length; k++) {
    const key = keys[ k ]
    const found = map[ key ];

    if (typeof found === 'undefined') {
      result.push(null)
    }
    else if (found instanceof Array) {
      let idx = order[ key ];

      if (typeof idx === 'undefined') {
        idx = 0
      }
      else {
        idx++
      }

      order[ key ] = idx;

      result.push(found[ idx ]);
    }
    else {
      result.push(found);
    }
  }

  return result;
};



export const processCursor = (entity, cursor, orderBy, reverse) => {

  const $LT = reverse ? '$gt' : '$lt'
  const $GT = reverse ? '$lt' : '$gt'
  const $LTE = reverse ? '$gte' : '$lte'
  const $GTE = reverse ? '$lte' : '$gte'

  const where = {}

  if (cursor) {
    passOrThrow(
      isArray(cursor[ entity.name ]),
      () => 'Incompatible cursor for this entity'
    )

    passOrThrow(
      isArray(orderBy),
      () => 'orderBy needs to be an array of order definitions'
    )

    const orderList = []

    const orderMap = {}
    orderBy.map(({attribute, direction}) => {

      passOrThrow(
        attribute && direction,
        () => 'orderBy needs to be an array of attributes and respective sort order'
      )

      orderMap[ attribute ] = direction
    })

    const primaryAttribute = entity.getPrimaryAttribute()
    const attributes = entity.getAttributes()

    let foundUniqueAttribute = false

    cursor[ entity.name ].map(filter => {
      if (filter.length !== 2) {
        throw new Error('Cursor malformed')
      }

      const attributeName = filter[0]

      passOrThrow(
        attributes[ attributeName ],
        () => `Unknown attribute '${attributeName}' used in cursor`
      )

      if (attributeName !== primaryAttribute.name) {
        passOrThrow(
          orderMap[ attributeName ],
          () => `Cursor works only on sorted attributes (check: '${attributeName}')`
        )
      }

      const attribute = attributes[ attributeName ]

      // limit where clause to the first attribute which is defined as unique
      if (!foundUniqueAttribute) {
        orderList.push(attributeName)
      }

      if (attribute.isUnique) {
        foundUniqueAttribute = true
      }

    })

    passOrThrow(
      foundUniqueAttribute,
      () => 'Cursor needs to have at least one attribute defined as unique'
    )


    // if more than 2 attributes are used for the cursor take only the first and the last (primary key)
    if (orderList.length > 2) {
      orderList.splice(1, (orderList.length - 2) )
    }


    // simple filter for single attributes
    if (orderList.length === 1) {

      const attributeName = orderList[0]
      const value = cursor[ entity.name ][0][1]

      if (orderMap[ attributeName ] === 'DESC') {
        where[ attributeName ] = {
          [ $LT ]: value
        }
      }
      else {
        where[ attributeName ] = {
          [ $GT ]: value
        }
      }

    }
    else {
      where.$not = {}

      cursor[ entity.name ].map(filter => {
        const attributeName = filter[0]
        const value = filter[1]

        // ignore attributes that obsolete due to prior unique attribute
        if (orderList.indexOf(attributeName) === -1) {
          return
        }

        const attribute = attributes[ attributeName ]

        if (attribute.isUnique) {
          if (orderMap[ attributeName ] === 'DESC') {
            where.$not[ attributeName ] = {
              [ $GTE ]: value
            }
          }
          else {
            where.$not[ attributeName ] = {
              [ $LTE ]: value
            }
          }
        }
        else {
          where.$not[ attributeName ] = value

          if (orderMap[ attributeName ] === 'DESC') {
            where[ attributeName ] = {
              [ $LTE ]: value
            }
          }
          else {
            where[ attributeName ] = {
              [ $GTE ]: value
            }
          }
        }

      })
    }
  }

  return where
}


export const processCursors = (entity, args) => {

  const {
    after,
    before,
  } = args;

  const where = {
    $and: [
      processCursor(entity, after, args.orderBy),
      processCursor(entity, before, args.orderBy, true),
    ]
  }

  return where
}



export const reverseString = (str) => {
  return str
    .split('')
    .reverse()
    .join('');
}



export const randomJson = () => {
  const ret = {}
  const count = casual.integer(1, 5)

  _.times(count, () => {
    const key = _.camelCase(casual.words(2))
    const value = Math.random() > 0.5
      ? casual.title
      : casual.integer()

    ret[key] = value
  })

  return ret
}
