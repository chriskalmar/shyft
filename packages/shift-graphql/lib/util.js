
import { constants } from 'shift-engine';
import _ from 'lodash';


// generate a type name from a model
export function generateTypeName(entityModel) {

  if (!entityModel || !_.isObject(entityModel)) {
    throw new Error('generateTypeName() entityModel needs to be defined')
  }

  if (!entityModel.name || !entityModel.domain) {
    throw new Error('generateTypeName() entityModel needs a name and a domain')
  }

  let name = entityModel.domain

  // if it's not the local provider use it for the node name (unique naming)
  if (constants.localProviderName !== entityModel.provider) {
    name = `${entityModel.provider}_${name}`
  }

  // add entity name
  name += `_${entityModel.name}`

  return _.camelCase(name)
}


export function upperCaseFirst(str) {
  return str.charAt(0).toUpperCase() + str.slice(1);
}



export function generateTypeNameUpperCase(entityModel) {
  return upperCaseFirst(
    generateTypeName(entityModel)
  )
}


export default {
  generateTypeName,
  upperCaseFirst,
  generateTypeNameUpperCase,
}
