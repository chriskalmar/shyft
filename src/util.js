import crypto from 'crypto';

// postgres limitation
const SYSTEM_NAME_MAX_LENGTH = 63;

export const generateIndexName = (entityName, attributes, suffix = '_idx') => {
  let ret = `${entityName}_${suffix}_`;
  const attributesChain = (attributes || []).join('_');
  const attributesSortedChain = (attributes || []).sort().join('_');

  if (ret.length + attributesChain.length > SYSTEM_NAME_MAX_LENGTH) {
    const hashLength = 10;
    const hash = crypto
      .createHash('sha256')
      .update(attributesSortedChain)
      .digest('hex')
      .substr(0, hashLength);

    const shortendAttributesChain = attributesChain.substr(
      0,
      SYSTEM_NAME_MAX_LENGTH - hashLength - ret.length - 1,
    );

    ret = `${entityName}_${shortendAttributesChain}_${hash}_${suffix}`;
  } else {
    ret = `${entityName}_${attributesChain}_${suffix}`;
  }

  return ret;
};

export const invertDirection = direction => {
  return direction === 'DESC' ? 'ASC' : 'DESC';
};

export const getLimit = (first, last) => {
  if (first >= 0) {
    return first;
  } else if (last >= 0) {
    return last;
  }

  return 10;
};

export const asyncForEach = async (array, callback) => {
  for (let index = 0; index < array.length; index++) {
    await callback(array[index], index, array);
  }
};

export const quote = item => {
  if (!item) {
    throw new Error('quote() requires an input');
  }

  // special care for json pointers
  if (item.includes('->')) {
    const arr = item.split('->');
    arr[0] = `"${arr[0]}"`;
    return arr.join('->');
  }

  return `"${item}"`;
};
