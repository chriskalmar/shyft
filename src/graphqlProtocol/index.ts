import {
  getTypeForEntityFromGraphRegistry,
  extendModelsForGql,
  registerActions,
  generateGraphQLSchema,
} from './generator';
import {
  GraphQLJSON,
  GraphQLCursor,
  GraphQLBigInt,
  GraphQLDateTime,
  GraphQLDate,
} from './dataTypes';
import { ProtocolGraphQL } from './ProtocolGraphQL';
import {
  RELAY_TYPE_PROMOTER_FIELD,
  MAX_PAGE_SIZE,
} from './protocolGraphqlConstants';
import {
  ProtocolGraphQLConfiguration,
  isProtocolGraphQLConfiguration,
} from './ProtocolGraphQLConfiguration';
import { fromBase64, toBase64 } from './util';

export {
  getTypeForEntityFromGraphRegistry,
  extendModelsForGql,
  registerActions,
  generateGraphQLSchema,
  GraphQLJSON,
  GraphQLCursor,
  GraphQLBigInt,
  GraphQLDateTime,
  GraphQLDate,
  ProtocolGraphQL,
  RELAY_TYPE_PROMOTER_FIELD,
  MAX_PAGE_SIZE,
  ProtocolGraphQLConfiguration,
  isProtocolGraphQLConfiguration,
  fromBase64,
  toBase64,
};
