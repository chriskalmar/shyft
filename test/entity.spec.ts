import './setupAndTearDown';
import { testGraphql } from './db';
import { asAdmin, asUser } from './testUtils';
import { gql } from '../src/graphqlProtocol/util';

describe('entity', () => {
  describe('mutation', () => {
    it('should accept inputs', async () => {
      const result = await testGraphql(
        gql`
          mutation saveWebsite {
            saveWebsite(
              input: {
                website: {
                  url: "https://thebestplacetocode.dev/"
                  category: CODING
                }
              }
            ) {
              website {
                id
                url
                category
                isSecure
                createdBy
                profileByCreatedBy {
                  id
                  username
                }
              }
            }
          }
        `,
        asUser(95),
      );

      expect(result).toMatchSnapshot();
    });

    const websiteCache = {};

    it('should use defaults', async () => {
      const result = await testGraphql(
        gql`
          mutation saveWebsite {
            saveWebsite(
              input: { website: { url: "https://example.com/path/index.html" } }
            ) {
              website {
                id
                nodeId
                url
                category
                isSecure
                createdBy
                profileByCreatedBy {
                  id
                  username
                }
              }
            }
          }
        `,
        asUser(98),
      );

      websiteCache['example-98'] = result.data.saveWebsite.website;

      expect(result).toMatchSnapshot();
    });

    it('should work with zero-attribute mutations', async () => {
      const result = await testGraphql(
        gql`
          mutation visitWebsiteById($id: ID!) {
            visitWebsiteById(input: { id: $id }) {
              website {
                id
                url
                category
                isSecure
                createdBy
                visitCount
                profileByCreatedBy {
                  id
                  username
                }
              }
            }
          }
        `,
        asUser(98),
        {
          id: websiteCache['example-98'].id,
        },
      );

      expect(result).toMatchSnapshot();
    });

    it('should work a pre processor', async () => {
      const result = await testGraphql(
        gql`
          mutation saveWebsite {
            saveWebsite(input: { website: { url: "I'm feeling lucky!" } }) {
              website {
                id
                nodeId
                url
                category
                isSecure
                createdBy
                profileByCreatedBy {
                  id
                  username
                }
              }
            }
          }
        `,
        asUser(98),
      );

      expect(result).toMatchSnapshot();
    });

    it('should use validators', async () => {
      const result = await testGraphql(
        gql`
          mutation saveWebsite {
            saveWebsite(
              input: { website: { url: "ftp://whatever.example.com" } }
            ) {
              website {
                id
                nodeId
                url
                category
                isSecure
                createdBy
                profileByCreatedBy {
                  id
                  username
                }
              }
            }
          }
        `,
        asUser(98),
      );

      expect(result).toMatchSnapshot();
    });

    it('should work with nodeId', async () => {
      const result = await testGraphql(
        gql`
          mutation visitWebsite($nodeId: ID!) {
            visitWebsite(input: { nodeId: $nodeId }) {
              website {
                id
                url
                category
                isSecure
                createdBy
                visitCount
                profileByCreatedBy {
                  id
                  username
                }
              }
            }
          }
        `,
        asUser(98),
        {
          nodeId: websiteCache['example-98'].nodeId,
        },
      );

      expect(result).toMatchSnapshot();
    });

    it('should fail on incompatible nodeId', async () => {
      const result = await testGraphql(
        gql`
          mutation visitWebsite($nodeId: ID!) {
            visitWebsite(input: { nodeId: $nodeId }) {
              website {
                id
                url
                category
                isSecure
                createdBy
                visitCount
                profileByCreatedBy {
                  id
                  username
                }
              }
            }
          }
        `,
        asUser(98),
        {
          nodeId: '123',
        },
      );

      expect(result).toMatchSnapshot();
    });

    it('should run multiple mutations in sequence', async () => {
      const result = await testGraphql(
        gql`
          mutation createLanguage {
            en: createLanguage(
              input: {
                language: {
                  name: "English"
                  nativeName: "English"
                  isoCode: "en"
                }
              }
            ) {
              language {
                id
              }
            }
            de: createLanguage(
              input: {
                language: {
                  name: "German"
                  nativeName: "Deutsch"
                  isoCode: "de"
                }
              }
            ) {
              language {
                id
              }
            }
            pl: createLanguage(
              input: {
                language: {
                  name: "Polish"
                  nativeName: "Polski"
                  isoCode: "pl"
                }
              }
            ) {
              language {
                id
              }
            }
          }
        `,
        asAdmin(),
      );

      expect(result).toMatchSnapshot();
    });

    it('should work with nested inputs', async () => {
      const result = await testGraphql(
        gql`
          mutation createtag {
            readLater: createTagNested(
              input: {
                tag: {
                  name: "Read Later"
                  languageByUniqueIsoCode: { isoCode: "en" }
                }
              }
            ) {
              tag {
                id
              }
            }
            learning: createTagNested(
              input: {
                tag: {
                  name: "Learning"
                  languageByUniqueIsoCode: { isoCode: "en" }
                }
              }
            ) {
              tag {
                id
              }
            }
            lernen: createTagNested(
              input: {
                tag: {
                  name: "Lernen"
                  languageByUniqueIsoCode: { isoCode: "de" }
                }
              }
            ) {
              tag {
                id
              }
            }
            nauka: createTagNested(
              input: {
                tag: {
                  name: "Nauka"
                  languageByUniqueIsoCode: { isoCode: "pl" }
                }
              }
            ) {
              tag {
                id
              }
            }
            fun: createTagNested(
              input: {
                tag: { name: "Fun", languageByUniqueIsoCode: { isoCode: "en" } }
              }
            ) {
              tag {
                id
              }
            }
          }
        `,
        asAdmin(),
      );

      expect(result).toMatchSnapshot();
    });

    it('should work with deep nested inputs', async () => {
      const result = await testGraphql(
        gql`
          mutation createWebsiteTagNested {
            first: createWebsiteTagNested(
              input: {
                websiteTag: {
                  websiteByUniqueUrl: {
                    url: "https://example.com/path/index.html"
                  }
                  tagByUniqueNameAndLanguage: {
                    name: "Fun"
                    languageByUniqueIsoCode: { isoCode: "en" }
                  }
                }
              }
            ) {
              websiteTag {
                id
              }
            }

            second: createWebsiteTagNested(
              input: {
                websiteTag: {
                  websiteByUniqueUrl: { url: "https://thebestplacetocode.dev/" }
                  tagByUniqueNameAndLanguage: {
                    name: "Nauka"
                    languageByUniqueName: { name: "Polish" }
                  }
                }
              }
            ) {
              websiteTag {
                id
              }
            }
          }
        `,
        asAdmin(),
      );

      expect(result).toMatchSnapshot();
    });

    it("should fail if nested inputs can't be resolved", async () => {
      const result = await testGraphql(
        gql`
          mutation createWebsiteTagNested {
            first: createWebsiteTagNested(
              input: {
                websiteTag: {
                  websiteByUniqueUrl: {
                    url: "https://example.com/path/index.html"
                  }
                  tagByUniqueNameAndLanguage: {
                    name: "Fun"
                    languageByUniqueIsoCode: { isoCode: "Does not exist" }
                  }
                }
              }
            ) {
              websiteTag {
                id
              }
            }

            second: createWebsiteTagNested(
              input: {
                websiteTag: {
                  websiteByUniqueUrl: { url: "https://thebestplacetocode.dev/" }
                  tagByUniqueNameAndLanguage: {
                    name: "Does not exist"
                    languageByUniqueName: { name: "Polish" }
                  }
                }
              }
            ) {
              websiteTag {
                id
              }
            }
          }
        `,
        asAdmin(),
      );

      expect(result).toMatchSnapshot();
    });

    it('should fail on ambiguous nested inputs', async () => {
      const result = await testGraphql(
        gql`
          mutation createWebsiteTagNested {
            createWebsiteTagNested(
              input: {
                websiteTag: {
                  websiteByUniqueUrl: {
                    url: "https://example.com/path/index.html"
                  }
                  tagByUniqueNameAndLanguage: {
                    name: "Fun"
                    languageByUniqueIsoCode: { isoCode: "en" }
                    languageByUniqueName: { name: "English" }
                  }
                }
              }
            ) {
              websiteTag {
                id
              }
            }
          }
        `,
        asAdmin(),
      );

      expect(result).toMatchSnapshot();
    });
  });

  describe('query', () => {
    it('should resolve computed values', async () => {
      const result = await testGraphql(
        gql`
          query allWebsites {
            allWebsites {
              totalCount
              edges {
                node {
                  id
                  url
                  category
                  isSecure
                  createdBy
                  profileByCreatedBy {
                    id
                    username
                  }
                }
              }
            }
          }
        `,
        asAdmin(),
      );

      expect(result).toMatchSnapshot();
    });

    it('should deep resolve edges on the graph', async () => {
      const result = await testGraphql(
        gql`
          query allBoardMembers {
            allBoardMembers(first: 3) {
              edges {
                node {
                  id
                  inviter
                  invitee
                  board
                  boardByBoard {
                    id
                    name
                    boardMembersByBoard(
                      orderBy: [BOARD_ASC, INVITEE_ASC]
                      first: 4
                    ) {
                      totalCount
                      edges {
                        node {
                          profileByInvitee {
                            id
                            username
                            boardsByOwner {
                              totalCount
                              edges {
                                node {
                                  name
                                  owner
                                }
                              }
                            }
                          }
                        }
                      }
                    }
                  }
                }
              }
            }
          }
        `,
        asAdmin(),
      );

      expect(result).toMatchSnapshot();
    });
  });
});
