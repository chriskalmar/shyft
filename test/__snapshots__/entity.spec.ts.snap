// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`entity mutation should accept inputs 1`] = `
Object {
  "data": Object {
    "saveWebsite": Object {
      "website": Object {
        "category": "CODING",
        "createdBy": "95",
        "id": "1",
        "isSecure": true,
        "profileByCreatedBy": Object {
          "id": "95",
          "username": "joel356",
        },
        "url": "https://thebestplacetocode.dev/",
      },
    },
  },
}
`;

exports[`entity mutation should fail if nested inputs can't be resolved 1`] = `
Object {
  "data": Object {
    "first": null,
    "second": null,
  },
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "NestedInstanceNotFoundOrAccessDenied",
      },
      "locations": Array [
        Object {
          "column": 13,
          "line": 3,
        },
      ],
      "message": "Nested instance at path 'tagByUniqueNameAndLanguage.languageByUniqueIsoCode' not found or access denied",
      "meta": undefined,
      "originalError": [Error: Nested instance at path 'tagByUniqueNameAndLanguage.languageByUniqueIsoCode' not found or access denied],
      "path": Array [
        "first",
      ],
    },
    Object {
      "extensions": Object {
        "code": "NestedInstanceNotFoundOrAccessDenied",
      },
      "locations": Array [
        Object {
          "column": 13,
          "line": 21,
        },
      ],
      "message": "Nested instance at path 'tagByUniqueNameAndLanguage' not found or access denied",
      "meta": undefined,
      "originalError": [Error: Nested instance at path 'tagByUniqueNameAndLanguage' not found or access denied],
      "path": Array [
        "second",
      ],
    },
  ],
}
`;

exports[`entity mutation should fail on ambiguous nested inputs 1`] = `
Object {
  "data": Object {
    "createWebsiteTagNested": null,
  },
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "AmbiguousNestedInputError",
      },
      "locations": Array [
        Object {
          "column": 13,
          "line": 3,
        },
      ],
      "message": "Only one of these fields may be used: language, languageByUniqueName, languageByUniqueIsoCode",
      "meta": undefined,
      "originalError": [Error: Only one of these fields may be used: language, languageByUniqueName, languageByUniqueIsoCode],
      "path": Array [
        "createWebsiteTagNested",
      ],
    },
  ],
}
`;

exports[`entity mutation should fail on incompatible nodeId 1`] = `
Object {
  "data": Object {
    "visitWebsite": null,
  },
  "errors": Array [
    Object {
      "extensions": Object {
        "code": undefined,
      },
      "locations": Array [
        Object {
          "column": 13,
          "line": 3,
        },
      ],
      "message": "Incompatible nodeId used with this mutation",
      "meta": undefined,
      "originalError": [Error: Incompatible nodeId used with this mutation],
      "path": Array [
        "visitWebsite",
      ],
    },
  ],
}
`;

exports[`entity mutation should run multiple mutations in sequence 1`] = `
Object {
  "data": Object {
    "de": Object {
      "language": Object {
        "id": "2",
      },
    },
    "en": Object {
      "language": Object {
        "id": "1",
      },
    },
    "pl": Object {
      "language": Object {
        "id": "3",
      },
    },
  },
}
`;

exports[`entity mutation should use defaults 1`] = `
Object {
  "data": Object {
    "saveWebsite": Object {
      "website": Object {
        "category": "MISC",
        "createdBy": "98",
        "id": "2",
        "isSecure": true,
        "nodeId": "d2Vic2l0ZToy",
        "profileByCreatedBy": Object {
          "id": "98",
          "username": "elza510",
        },
        "url": "https://example.com/path/index.html",
      },
    },
  },
}
`;

exports[`entity mutation should use validators 1`] = `
Object {
  "data": Object {
    "saveWebsite": null,
  },
  "errors": Array [
    Object {
      "extensions": Object {
        "code": undefined,
      },
      "locations": Array [
        Object {
          "column": 13,
          "line": 3,
        },
      ],
      "message": "Unknown protocol, should be http:// or https://",
      "meta": undefined,
      "originalError": [Error: Unknown protocol, should be http:// or https://],
      "path": Array [
        "saveWebsite",
      ],
    },
  ],
}
`;

exports[`entity mutation should work a pre processor 1`] = `
Object {
  "data": Object {
    "saveWebsite": Object {
      "website": Object {
        "category": "NEWS",
        "createdBy": "98",
        "id": "3",
        "isSecure": false,
        "nodeId": "d2Vic2l0ZToz",
        "profileByCreatedBy": Object {
          "id": "98",
          "username": "elza510",
        },
        "url": "http://a.random.unsecured.news.website.com",
      },
    },
  },
}
`;

exports[`entity mutation should work with deep nested inputs 1`] = `
Object {
  "data": Object {
    "first": Object {
      "websiteTag": Object {
        "id": "1",
      },
    },
    "second": Object {
      "websiteTag": Object {
        "id": "2",
      },
    },
  },
}
`;

exports[`entity mutation should work with nested inputs 1`] = `
Object {
  "data": Object {
    "fun": Object {
      "tag": Object {
        "id": "5",
      },
    },
    "learning": Object {
      "tag": Object {
        "id": "2",
      },
    },
    "lernen": Object {
      "tag": Object {
        "id": "3",
      },
    },
    "nauka": Object {
      "tag": Object {
        "id": "4",
      },
    },
    "readLater": Object {
      "tag": Object {
        "id": "1",
      },
    },
  },
}
`;

exports[`entity mutation should work with nodeId 1`] = `
Object {
  "data": Object {
    "visitWebsite": Object {
      "website": Object {
        "category": "MISC",
        "createdBy": "98",
        "id": "2",
        "isSecure": true,
        "profileByCreatedBy": Object {
          "id": "98",
          "username": "elza510",
        },
        "url": "https://example.com/path/index.html",
        "visitCount": 2,
      },
    },
  },
}
`;

exports[`entity mutation should work with zero-attribute mutations 1`] = `
Object {
  "data": Object {
    "visitWebsiteById": Object {
      "website": Object {
        "category": "MISC",
        "createdBy": "98",
        "id": "2",
        "isSecure": true,
        "profileByCreatedBy": Object {
          "id": "98",
          "username": "elza510",
        },
        "url": "https://example.com/path/index.html",
        "visitCount": 1,
      },
    },
  },
}
`;

exports[`entity query should deep resolve edges on the graph 1`] = `
Object {
  "data": Object {
    "allBoardMembers": Object {
      "edges": Array [
        Object {
          "node": Object {
            "board": "41",
            "boardByBoard": Object {
              "boardMembersByBoard": Object {
                "edges": Array [
                  Object {
                    "node": Object {
                      "profileByInvitee": Object {
                        "boardsByOwner": Object {
                          "edges": Array [
                            Object {
                              "node": Object {
                                "name": "Sunt quo",
                                "owner": "10",
                              },
                            },
                          ],
                          "totalCount": 1,
                        },
                        "id": "10",
                        "username": "leon807",
                      },
                    },
                  },
                  Object {
                    "node": Object {
                      "profileByInvitee": Object {
                        "boardsByOwner": Object {
                          "edges": Array [
                            Object {
                              "node": Object {
                                "name": "Culpa facilis",
                                "owner": "20",
                              },
                            },
                          ],
                          "totalCount": 1,
                        },
                        "id": "20",
                        "username": "katrina560",
                      },
                    },
                  },
                  Object {
                    "node": Object {
                      "profileByInvitee": Object {
                        "boardsByOwner": Object {
                          "edges": Array [
                            Object {
                              "node": Object {
                                "name": "Nobis totam",
                                "owner": "41",
                              },
                            },
                          ],
                          "totalCount": 1,
                        },
                        "id": "41",
                        "username": "myrl112",
                      },
                    },
                  },
                  Object {
                    "node": Object {
                      "profileByInvitee": Object {
                        "boardsByOwner": Object {
                          "edges": Array [
                            Object {
                              "node": Object {
                                "name": "Illo sit quia",
                                "owner": "59",
                              },
                            },
                          ],
                          "totalCount": 1,
                        },
                        "id": "59",
                        "username": "kara119",
                      },
                    },
                  },
                ],
                "totalCount": 7,
              },
              "id": "41",
              "name": "Enim dolore",
            },
            "id": "1",
            "invitee": "95",
            "inviter": "95",
          },
        },
        Object {
          "node": Object {
            "board": "21",
            "boardByBoard": Object {
              "boardMembersByBoard": Object {
                "edges": Array [
                  Object {
                    "node": Object {
                      "profileByInvitee": Object {
                        "boardsByOwner": Object {
                          "edges": Array [
                            Object {
                              "node": Object {
                                "name": "Sunt quo",
                                "owner": "10",
                              },
                            },
                          ],
                          "totalCount": 1,
                        },
                        "id": "10",
                        "username": "leon807",
                      },
                    },
                  },
                  Object {
                    "node": Object {
                      "profileByInvitee": Object {
                        "boardsByOwner": Object {
                          "edges": Array [
                            Object {
                              "node": Object {
                                "name": "Aut molestiae",
                                "owner": "12",
                              },
                            },
                          ],
                          "totalCount": 1,
                        },
                        "id": "12",
                        "username": "emmet741",
                      },
                    },
                  },
                  Object {
                    "node": Object {
                      "profileByInvitee": Object {
                        "boardsByOwner": Object {
                          "edges": Array [
                            Object {
                              "node": Object {
                                "name": "Delectus qui maxime",
                                "owner": "42",
                              },
                            },
                            Object {
                              "node": Object {
                                "name": "Ut et",
                                "owner": "42",
                              },
                            },
                            Object {
                              "node": Object {
                                "name": "Voluptate tempora veritatis",
                                "owner": "42",
                              },
                            },
                          ],
                          "totalCount": 3,
                        },
                        "id": "42",
                        "username": "alva280",
                      },
                    },
                  },
                  Object {
                    "node": Object {
                      "profileByInvitee": Object {
                        "boardsByOwner": Object {
                          "edges": Array [],
                          "totalCount": 0,
                        },
                        "id": "45",
                        "username": "salvador975",
                      },
                    },
                  },
                ],
                "totalCount": 8,
              },
              "id": "21",
              "name": "Quia sapiente",
            },
            "id": "2",
            "invitee": "45",
            "inviter": "45",
          },
        },
        Object {
          "node": Object {
            "board": "4",
            "boardByBoard": Object {
              "boardMembersByBoard": Object {
                "edges": Array [
                  Object {
                    "node": Object {
                      "profileByInvitee": Object {
                        "boardsByOwner": Object {
                          "edges": Array [],
                          "totalCount": 0,
                        },
                        "id": "5",
                        "username": "griffin807",
                      },
                    },
                  },
                  Object {
                    "node": Object {
                      "profileByInvitee": Object {
                        "boardsByOwner": Object {
                          "edges": Array [
                            Object {
                              "node": Object {
                                "name": "Qui rem ut",
                                "owner": "9",
                              },
                            },
                          ],
                          "totalCount": 1,
                        },
                        "id": "9",
                        "username": "omari409",
                      },
                    },
                  },
                  Object {
                    "node": Object {
                      "profileByInvitee": Object {
                        "boardsByOwner": Object {
                          "edges": Array [
                            Object {
                              "node": Object {
                                "name": "Voluptatum necessitatibus molestias",
                                "owner": "17",
                              },
                            },
                          ],
                          "totalCount": 1,
                        },
                        "id": "17",
                        "username": "fred777",
                      },
                    },
                  },
                  Object {
                    "node": Object {
                      "profileByInvitee": Object {
                        "boardsByOwner": Object {
                          "edges": Array [
                            Object {
                              "node": Object {
                                "name": "Non iusto unde",
                                "owner": "30",
                              },
                            },
                          ],
                          "totalCount": 1,
                        },
                        "id": "30",
                        "username": "carissa722",
                      },
                    },
                  },
                ],
                "totalCount": 8,
              },
              "id": "4",
              "name": "Delectus qui maxime",
            },
            "id": "3",
            "invitee": "17",
            "inviter": "17",
          },
        },
      ],
    },
  },
}
`;

exports[`entity query should resolve computed values 1`] = `
Object {
  "data": Object {
    "allWebsites": Object {
      "edges": Array [
        Object {
          "node": Object {
            "category": "CODING",
            "createdBy": "95",
            "id": "1",
            "isSecure": true,
            "profileByCreatedBy": Object {
              "id": "95",
              "username": "joel356",
            },
            "url": "https://thebestplacetocode.dev/",
          },
        },
        Object {
          "node": Object {
            "category": "MISC",
            "createdBy": "98",
            "id": "2",
            "isSecure": true,
            "profileByCreatedBy": Object {
              "id": "98",
              "username": "elza510",
            },
            "url": "https://example.com/path/index.html",
          },
        },
        Object {
          "node": Object {
            "category": "NEWS",
            "createdBy": "98",
            "id": "3",
            "isSecure": false,
            "profileByCreatedBy": Object {
              "id": "98",
              "username": "elza510",
            },
            "url": "http://a.random.unsecured.news.website.com",
          },
        },
      ],
      "totalCount": 3,
    },
  },
}
`;
